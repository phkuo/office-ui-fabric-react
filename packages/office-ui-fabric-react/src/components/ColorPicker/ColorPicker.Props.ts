export interface IColorPickerProps {
  /**
   * The CSS-compatible string to describe the initially selected color
   */
  color: string;

  /**
   * Callback issued when the user changes the color
   */
  onColorChanged?: (color: string) => void;
}
