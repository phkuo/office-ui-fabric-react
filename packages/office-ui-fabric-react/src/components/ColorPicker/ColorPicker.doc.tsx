import * as React from 'react';
import { ColorPickerBasicExample } from './examples/ColorPicker.Basic.Example';

import { IDocPageProps } from '../../common/DocPage.types';
import { ColorPickerStatus } from './ColorPicker.checklist';

const ColorPickerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/examples/ColorPicker.Basic.Example.tsx') as string;
export const ColorPickerPageProps: IDocPageProps = {
  title: 'ColorPicker',
  componentName: 'ColorPicker',
  componentUrl:
    'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ColorPicker',
  componentStatus: ColorPickerStatus,
  examples: [
    {
      title: 'Default ColorPicker',
      code: ColorPickerBasicExampleCode,
      view: <ColorPickerBasicExample />
    }
  ],
  propertiesTablesSources: [
    require<string>('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorPicker.types.ts'),
    require<string>('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorRectangle/ColorRectangle.types.ts'),
    require<string>('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorSlider/ColorSlider.types.ts')
  ],
  overview: require<string>('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/docs/ColorPickerOverview.md'),
  bestPractices: '',
  dos: '',
  donts: '',
  isHeaderVisible: true,
  isFeedbackVisible: true
};
