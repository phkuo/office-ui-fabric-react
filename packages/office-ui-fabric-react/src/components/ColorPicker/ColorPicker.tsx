import * as React from 'react';
import { IColorPickerProps } from './ColorPicker.Props';
import { TextField } from '../../TextField';
import { ColorRectangle } from './ColorRectangle';
import { ColorSlider } from './ColorSlider';
import { autobind } from '../../utilities/autobind';
import {
  getColorFromString,
  getColorFromRGBA,
  updateA,
  updateH,
  updateSV
} from '../../utilities/Color/Colors';
import {
  IColor,
  MAX_COLOR_HUE
} from '../../utilities/Color/IColor';
import './ColorPicker.scss';

export interface IColorPickerState {
  isOpen: boolean;
  color: IColor;
}

export interface IColor {
  r: number;
  g: number;
  b: number;
  a: number;
  h: number;
  s: number;
  v: number;
  hex: string;
  str: string;
}

export class ColorPicker extends React.Component<IColorPickerProps, IColorPickerState> {
  private hexText: TextField;
  private rText: TextField;
  private gText: TextField;
  private bText: TextField;
  private aText: TextField;

  constructor(props: IColorPickerProps) {
    super(props);

    this.state = {
      color: getColorFromString(props.color)
    } as IColorPickerState;
  }

  public componentWillReceiveProps(newProps: IColorPickerProps) {
    if (newProps.color) {
      this._updateColor(getColorFromString(newProps.color));
    }
  }

  public render() {
    let { color } = this.state;

    return (
      <div className='ms-ColorPicker'>
        <div className='ms-ColorPicker-panel'>
          <ColorRectangle color={ color } onSVChanged={ this._onSVChanged } />
          <ColorSlider
            className='is-hue'
            minValue={ 0 }
            maxValue={ MAX_COLOR_HUE }
            initialValue={ color.h }
            onChanged={ this._onHChanged }
          />
          <ColorSlider
            className='is-alpha'
            overlayStyle={ { background: `linear-gradient(to right, transparent 0, ${color.str} 100%)` } }
            minValue={ 0 }
            maxValue={ 100 }
            initialValue={ color.a }
            onChanged={ this._onAChanged }
          />
          <table className='ms-ColorPicker-table' cellPadding='0' cellSpacing='0'>
            <thead>
              <tr className='ms-font-s'>
                <td>Hex</td>
                <td>Red</td>
                <td>Green</td>
                <td>Blue</td>
                <td>Alpha</td>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td><TextField className='ms-ColorPicker-input' value={ color.hex } ref={ (ref) => this.hexText = ref } onBlur={ this._onHexChanged } /></td>
                <td style={ { width: '18%' } }><TextField className='ms-ColorPicker-input' onBlur={ this._onRGBAChanged }
                  value={ String(color.r) } ref={ (ref) => this.rText = ref } /></td>
                <td style={ { width: '18%' } }><TextField className='ms-ColorPicker-input' onBlur={ this._onRGBAChanged }
                  value={ String(color.g) } ref={ (ref) => this.gText = ref } /></td>
                <td style={ { width: '18%' } }><TextField className='ms-ColorPicker-input' onBlur={ this._onRGBAChanged }
                  value={ String(color.b) } ref={ (ref) => this.bText = ref } /></td>
                <td style={ { width: '18%' } }><TextField className='ms-ColorPicker-input' onBlur={ this._onRGBAChanged }
                  value={ String(color.a) } ref={ (ref) => this.aText = ref } /></td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    );
  }

  @autobind
  private _onSVChanged(s: number, v: number) {
    this._updateColor(updateSV(this.state.color, s, v));
  }

  @autobind
  private _onHChanged(h: number) {
    this._updateColor(updateH(this.state.color, h));
  }

  @autobind
  private _onAChanged(a: number) {
    this._updateColor(updateA(this.state.color, a));
  }

  @autobind
  private _onHexChanged() {
    this._updateColor(getColorFromString("#" + this.hexText.value));
  }

  @autobind
  private _onRGBAChanged() {
    this._updateColor(getColorFromRGBA({
      r: Number(this.rText.value),
      g: Number(this.gText.value),
      b: Number(this.bText.value),
      a: Number(this.aText.value)
    }));
  }

  private _updateColor(newColor: IColor) {
    let { onColorChanged } = this.props;

    if (newColor.str !== this.state.color.str) {
      this.setState({
        color: newColor
      } as IColorPickerState, () => {
        if (onColorChanged) {
          onColorChanged(newColor.str);
        }
      });
    }
  }
}
