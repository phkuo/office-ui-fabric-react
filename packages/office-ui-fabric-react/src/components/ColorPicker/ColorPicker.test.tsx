/* tslint:disable:no-unused-variable */
import * as React from 'react';
/* tslint:enable:no-unused-variable */
import * as renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { ColorPicker } from './ColorPicker';
import { ColorPickerBase, IColorPickerState } from './ColorPicker.base';
import { IColorPicker, IColorPickerProps } from './ColorPicker.types';
import { IColor, getColorFromString } from '../../utilities/color/colors';

describe('ColorPicker', () => {
  it('renders ColorPicker correctly', () => {
    const component = renderer.create(<ColorPicker color="#FFFFFF" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Props are correctly parsed', () => {
    let colorPickerComponent: any;
    const setRef = (ref: IColorPicker): void => {
      colorPickerComponent = ref;
    };

    mount(<ColorPicker color="#FFFFFF" componentRef={setRef} />);

    expect(colorPickerComponent.state.color.hex).toEqual('ffffff');
  });

  it('Reacts to props change', () => {
    let colorPickerComponent: any;
    const setRef = (ref: IColorPicker): void => {
      colorPickerComponent = ref;
    };

    mount(<ColorPicker color="#FFFFFF" componentRef={setRef} />);

    const component = colorPickerComponent;
    colorPickerComponent.componentWillReceiveProps({ color: '#AEAEAE' });
    expect(component.state.color.hex).toEqual('aeaeae');
  });

  it('onColorChange is called', () => {
    let color = '#FFFFFF';
    let newColorObject;
    const onColorChanged = (str: string, colorObject: IColor): void => {
      color = str;
      newColorObject = colorObject;
    };

    let colorPickerComponent: any;
    const setRef = (ref: IColorPicker): void => {
      colorPickerComponent = ref;
    };

    mount(<ColorPicker color={color} componentRef={setRef} onColorChanged={onColorChanged} />);

    const newColor = '#AEAEAE';
    const component = colorPickerComponent;
    colorPickerComponent.componentWillReceiveProps({ color: newColor });

    expect(component.state.color.hex).toEqual('aeaeae');
    expect(color).toEqual(newColor);
    expect(newColorObject).toEqual(getColorFromString(newColor));
  });

  it('Hides alpha control slider', () => {
    const wrapper = mount(<ColorPicker color="#FFFFFF" alphaSliderHidden={true} />);

    const alphaSlider = wrapper.find('.is-alpha');
    const tableHeaders = wrapper.find('.ms-ColorPicker-table > thead > tr > td');
    const tableInputs = wrapper.find('.ms-ColorPicker-table > tbody> tr > td');

    // There should only be table headers and inputs for hex, red, green, and blue (no alpha)
    const expectedTableComponents = 4;

    expect(alphaSlider.exists()).toBe(false);
    expect(tableHeaders).toHaveLength(expectedTableComponents);
    expect(tableInputs).toHaveLength(expectedTableComponents);
  });

  it('Renders default RGBA/Hex strings', () => {
    const wrapper = mount(<ColorPicker color="#FFFFFF" />);

    const tableHeaders = wrapper.find('.ms-ColorPicker-table > thead > tr > td');
    const textHeaders = [
      ColorPickerBase.defaultProps.hexLabel,
      ColorPickerBase.defaultProps.redLabel,
      ColorPickerBase.defaultProps.greenLabel,
      ColorPickerBase.defaultProps.blueLabel,
      ColorPickerBase.defaultProps.alphaLabel
    ];

    tableHeaders.forEach((node, index) => {
      expect(node.text()).toEqual(textHeaders[index]);
    });
  });

  it('Renders custom RGBA/Hex strings', () => {
    const textHeaders = ['Custom Hex', 'Custom Red', 'Custom Green', 'Custom Blue', 'Custom Alpha'];

    const wrapper = mount(
      <ColorPicker
        color="#FFFFFF"
        hexLabel={textHeaders[0]}
        redLabel={textHeaders[1]}
        greenLabel={textHeaders[2]}
        blueLabel={textHeaders[3]}
        alphaLabel={textHeaders[4]}
      />
    );

    const tableHeaders = wrapper.find('.ms-ColorPicker-table > thead > tr > td');
    tableHeaders.forEach((node, index) => {
      expect(node.text()).toEqual(textHeaders[index]);
    });
  });

  it('Keeps color value when tabbing between Hex and RGBA text inputs', () => {
    const colorStringValue = 'ffffff';
    let colorChangeCalled = false;
    const onColorChanged = (newColor: string): void => {
      colorChangeCalled = newColor !== colorStringValue;
    };

    let colorPickerComponent: any;
    const setRef = (ref: IColorPicker): void => {
      colorPickerComponent = ref;
    };

    const inputClassName = 'input-tab-test';
    const wrapper = mount<IColorPickerProps, IColorPickerState>(
      <ColorPicker
        color={`#${colorStringValue}`}
        onColorChanged={onColorChanged}
        componentRef={setRef}
        styles={{ input: inputClassName }}
      />
    );

    expect(colorPickerComponent.state.color.hex).toEqual(colorStringValue);
    expect(colorChangeCalled).toBeFalsy();

    // Tab between text inputs checking state after each time.
    const allInputs = wrapper.find(`.${inputClassName} input`);
    expect(allInputs.length).toBe(5);

    allInputs.forEach(input => {
      input.simulate('blur');

      expect(colorPickerComponent.state.color.hex).toEqual(colorStringValue);
      expect(colorChangeCalled).toBeFalsy();
    });
  });
});
