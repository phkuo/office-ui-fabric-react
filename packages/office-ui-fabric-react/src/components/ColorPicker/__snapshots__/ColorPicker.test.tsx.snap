// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker renders ColorPicker correctly 1`] = `
<div
  className=
      ms-ColorPicker
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        max-width: 300px;
        position: relative;
      }
>
  <div
    className=
        ms-ColorPicker-panel
        {
          padding-bottom: 16px;
          padding-left: 16px;
          padding-right: 16px;
          padding-top: 16px;
        }
  >
    <div
      className=
          ms-ColorPicker-colorRect
          {
            margin-bottom: 10px;
            position: relative;
          }
          @media screen and (-ms-high-contrast: active){& {
            -ms-high-contrast-adjust: none;
          }
      onMouseDown={[Function]}
      style={
        Object {
          "backgroundColor": "#ff0000",
          "minHeight": 220,
          "minWidth": 220,
        }
      }
    >
      <div
        className=
            ms-ColorPicker-light
            {
              background: linear-gradient(to right, white 0%, transparent 100%);
              bottom: 0px;
              left: 0px;
              position: absolute;
              right: 0px;
              top: 0px;
            }
      />
      <div
        className=
            ms-ColorPicker-dark
            {
              background: linear-gradient(to bottom, transparent 0, #000 100%);
              bottom: 0px;
              left: 0px;
              position: absolute;
              right: 0px;
              top: 0px;
            }
      />
      <div
        className=
            ms-ColorPicker-thumb
            {
              background: white;
              border-radius: 50%;
              border: 1px solid rgba(255,255,255,.8);
              box-shadow: 0 0 15px -5px black;
              height: 20px;
              position: absolute;
              transform: translate(-50%, -50%);
              width: 20px;
            }
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "left": "0%",
            "top": "0%",
          }
        }
      />
    </div>
    <div
      className=
          ms-ColorPicker-slider
          is-hue
          {
            border: 1px solid #eaeaea;
            box-sizing: border-box;
            height: 20px;
            margin-bottom: 5px;
            position: relative;
          }
      onMouseDown={[Function]}
      style={
        Object {
          "background": "linear-gradient(to left,red 0,#f09 10%,#cd00ff 20%,#3200ff 30%,#06f 40%,#00fffd 50%,#0f6 60%,#35ff00 70%,#cdff00 80%,#f90 90%,red 100%)",
        }
      }
    >
      <div
        className=
            ms-ColorPicker-sliderOverlay
            {
              bottom: 0px;
              content: ;
              left: 0px;
              position: absolute;
              right: 0px;
              top: 0px;
            }
      />
      <div
        className=
            ms-ColorPicker-thumb
            is-slider
            {
              background: white;
              border-radius: 50%;
              border: 1px solid rgba(255,255,255,.8);
              box-shadow: 0 0 15px -5px black;
              height: 20px;
              position: absolute;
              top: 50%;
              transform: translate(-50%, -50%);
              width: 20px;
            }
        style={
          Object {
            "left": "0%",
          }
        }
      />
    </div>
    <div
      className=
          ms-ColorPicker-slider
          is-alpha
          {
            border: 1px solid #eaeaea;
            box-sizing: border-box;
            height: 20px;
            margin-bottom: 5px;
            position: relative;
          }
      onMouseDown={[Function]}
      style={
        Object {
          "backgroundImage": "url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAJUlEQVQYV2N89erVfwY0ICYmxoguxjgUFKI7GsTH5m4M3w1ChQC1/Ca8i2n1WgAAAABJRU5ErkJggg==)",
        }
      }
    >
      <div
        className=
            ms-ColorPicker-sliderOverlay
            {
              bottom: 0px;
              content: ;
              left: 0px;
              position: absolute;
              right: 0px;
              top: 0px;
            }
        style={
          Object {
            "background": "linear-gradient(to right, transparent 0, #ffffff 100%)",
          }
        }
      />
      <div
        className=
            ms-ColorPicker-thumb
            is-slider
            {
              background: white;
              border-radius: 50%;
              border: 1px solid rgba(255,255,255,.8);
              box-shadow: 0 0 15px -5px black;
              height: 20px;
              position: absolute;
              top: 50%;
              transform: translate(-50%, -50%);
              width: 20px;
            }
        style={
          Object {
            "left": "100%",
          }
        }
      />
    </div>
    <table
      cellPadding="0"
      cellSpacing="0"
      className=
          ms-ColorPicker-table
          {
            table-layout: fixed;
            width: 100%;
          }
    >
      <thead>
        <tr
          className=

              {
                -moz-osx-font-smoothing: grayscale;
                -webkit-font-smoothing: antialiased;
                font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                font-size: 12px;
                font-weight: 400;
              }
        >
          <td
            className=

                {
                  width: 25%;
                }
          >
            Hex
          </td>
          <td>
            Red
          </td>
          <td>
            Green
          </td>
          <td>
            Blue
          </td>
          <td>
            Alpha
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div
              className=
                  ms-TextField
                  ms-ColorPicker-input
                  {
                    border: none;
                    box-shadow: none;
                    box-sizing: border-box;
                    height: 30px;
                    margin-bottom: 0px;
                    margin-left: 0px;
                    margin-right: 0px;
                    margin-top: 0px;
                    padding-bottom: 0px;
                    padding-left: 0px;
                    padding-right: 0px;
                    padding-top: 0px;
                    position: relative;
                    width: 100%;
                  }
                  @media screen and (-ms-high-contrast: active){& {
                    border-width: 2px;
                  }
                  &.ms-TextField {
                    padding-right: 2px;
                  }
                  & .ms-TextField-field {
                    min-width: auto;
                    padding-bottom: 5px;
                    padding-left: 5px;
                    padding-right: 5px;
                    padding-top: 5px;
                    text-overflow: clip;
                  }
            >
              <div
                className=
                    ms-TextField-wrapper

              >
                <div
                  className=
                      ms-TextField-fieldGroup
                      {
                        align-items: stretch;
                        background: #ffffff;
                        border: 1px solid #a6a6a6;
                        box-shadow: none;
                        box-sizing: border-box;
                        display: flex;
                        flex-direction: row;
                        height: 32px;
                        margin-bottom: 0px;
                        margin-left: 0px;
                        margin-right: 0px;
                        margin-top: 0px;
                        padding-bottom: 0px;
                        padding-left: 0px;
                        padding-right: 0px;
                        padding-top: 0px;
                        position: relative;
                      }
                      &:hover {
                        border-color: #333333;
                      }
                      @media screen and (-ms-high-contrast: active){&:hover {
                        border-color: Highlight;
                      }
                >
                  <input
                    aria-invalid={false}
                    aria-label="Hex"
                    className=
                        ms-TextField-field
                        {
                          background-color: transparent;
                          background: none;
                          border-radius: 0px;
                          border: none;
                          box-shadow: none;
                          box-sizing: border-box;
                          color: #333333;
                          font-size: 14px;
                          margin-bottom: 0px;
                          margin-left: 0px;
                          margin-right: 0px;
                          margin-top: 0px;
                          min-width: 0px;
                          outline: 0px;
                          padding-bottom: 0;
                          padding-left: 12px;
                          padding-right: 12px;
                          padding-top: 0;
                          text-overflow: ellipsis;
                          width: 100%;
                        }
                        &:active, &:focus, &:hover {
                          outline: 0px;
                        }
                        &::-ms-clear {
                          display: none;
                        }
                        &::placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                        &:-ms-input-placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                    id="TextField0"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    spellCheck={false}
                    type="text"
                    value="ffffff"
                  />
                </div>
              </div>
            </div>
          </td>
          <td
            style={
              Object {
                "width": "18%",
              }
            }
          >
            <div
              className=
                  ms-TextField
                  ms-ColorPicker-input
                  {
                    border: none;
                    box-shadow: none;
                    box-sizing: border-box;
                    height: 30px;
                    margin-bottom: 0px;
                    margin-left: 0px;
                    margin-right: 0px;
                    margin-top: 0px;
                    padding-bottom: 0px;
                    padding-left: 0px;
                    padding-right: 0px;
                    padding-top: 0px;
                    position: relative;
                    width: 100%;
                  }
                  @media screen and (-ms-high-contrast: active){& {
                    border-width: 2px;
                  }
                  &.ms-TextField {
                    padding-right: 2px;
                  }
                  & .ms-TextField-field {
                    min-width: auto;
                    padding-bottom: 5px;
                    padding-left: 5px;
                    padding-right: 5px;
                    padding-top: 5px;
                    text-overflow: clip;
                  }
            >
              <div
                className=
                    ms-TextField-wrapper

              >
                <div
                  className=
                      ms-TextField-fieldGroup
                      {
                        align-items: stretch;
                        background: #ffffff;
                        border: 1px solid #a6a6a6;
                        box-shadow: none;
                        box-sizing: border-box;
                        display: flex;
                        flex-direction: row;
                        height: 32px;
                        margin-bottom: 0px;
                        margin-left: 0px;
                        margin-right: 0px;
                        margin-top: 0px;
                        padding-bottom: 0px;
                        padding-left: 0px;
                        padding-right: 0px;
                        padding-top: 0px;
                        position: relative;
                      }
                      &:hover {
                        border-color: #333333;
                      }
                      @media screen and (-ms-high-contrast: active){&:hover {
                        border-color: Highlight;
                      }
                >
                  <input
                    aria-invalid={false}
                    aria-label="Red"
                    className=
                        ms-TextField-field
                        {
                          background-color: transparent;
                          background: none;
                          border-radius: 0px;
                          border: none;
                          box-shadow: none;
                          box-sizing: border-box;
                          color: #333333;
                          font-size: 14px;
                          margin-bottom: 0px;
                          margin-left: 0px;
                          margin-right: 0px;
                          margin-top: 0px;
                          min-width: 0px;
                          outline: 0px;
                          padding-bottom: 0;
                          padding-left: 12px;
                          padding-right: 12px;
                          padding-top: 0;
                          text-overflow: ellipsis;
                          width: 100%;
                        }
                        &:active, &:focus, &:hover {
                          outline: 0px;
                        }
                        &::-ms-clear {
                          display: none;
                        }
                        &::placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                        &:-ms-input-placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                    id="TextField2"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    spellCheck={false}
                    type="text"
                    value="255"
                  />
                </div>
              </div>
            </div>
          </td>
          <td
            style={
              Object {
                "width": "18%",
              }
            }
          >
            <div
              className=
                  ms-TextField
                  ms-ColorPicker-input
                  {
                    border: none;
                    box-shadow: none;
                    box-sizing: border-box;
                    height: 30px;
                    margin-bottom: 0px;
                    margin-left: 0px;
                    margin-right: 0px;
                    margin-top: 0px;
                    padding-bottom: 0px;
                    padding-left: 0px;
                    padding-right: 0px;
                    padding-top: 0px;
                    position: relative;
                    width: 100%;
                  }
                  @media screen and (-ms-high-contrast: active){& {
                    border-width: 2px;
                  }
                  &.ms-TextField {
                    padding-right: 2px;
                  }
                  & .ms-TextField-field {
                    min-width: auto;
                    padding-bottom: 5px;
                    padding-left: 5px;
                    padding-right: 5px;
                    padding-top: 5px;
                    text-overflow: clip;
                  }
            >
              <div
                className=
                    ms-TextField-wrapper

              >
                <div
                  className=
                      ms-TextField-fieldGroup
                      {
                        align-items: stretch;
                        background: #ffffff;
                        border: 1px solid #a6a6a6;
                        box-shadow: none;
                        box-sizing: border-box;
                        display: flex;
                        flex-direction: row;
                        height: 32px;
                        margin-bottom: 0px;
                        margin-left: 0px;
                        margin-right: 0px;
                        margin-top: 0px;
                        padding-bottom: 0px;
                        padding-left: 0px;
                        padding-right: 0px;
                        padding-top: 0px;
                        position: relative;
                      }
                      &:hover {
                        border-color: #333333;
                      }
                      @media screen and (-ms-high-contrast: active){&:hover {
                        border-color: Highlight;
                      }
                >
                  <input
                    aria-invalid={false}
                    aria-label="Green"
                    className=
                        ms-TextField-field
                        {
                          background-color: transparent;
                          background: none;
                          border-radius: 0px;
                          border: none;
                          box-shadow: none;
                          box-sizing: border-box;
                          color: #333333;
                          font-size: 14px;
                          margin-bottom: 0px;
                          margin-left: 0px;
                          margin-right: 0px;
                          margin-top: 0px;
                          min-width: 0px;
                          outline: 0px;
                          padding-bottom: 0;
                          padding-left: 12px;
                          padding-right: 12px;
                          padding-top: 0;
                          text-overflow: ellipsis;
                          width: 100%;
                        }
                        &:active, &:focus, &:hover {
                          outline: 0px;
                        }
                        &::-ms-clear {
                          display: none;
                        }
                        &::placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                        &:-ms-input-placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                    id="TextField4"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    spellCheck={false}
                    type="text"
                    value="255"
                  />
                </div>
              </div>
            </div>
          </td>
          <td
            style={
              Object {
                "width": "18%",
              }
            }
          >
            <div
              className=
                  ms-TextField
                  ms-ColorPicker-input
                  {
                    border: none;
                    box-shadow: none;
                    box-sizing: border-box;
                    height: 30px;
                    margin-bottom: 0px;
                    margin-left: 0px;
                    margin-right: 0px;
                    margin-top: 0px;
                    padding-bottom: 0px;
                    padding-left: 0px;
                    padding-right: 0px;
                    padding-top: 0px;
                    position: relative;
                    width: 100%;
                  }
                  @media screen and (-ms-high-contrast: active){& {
                    border-width: 2px;
                  }
                  &.ms-TextField {
                    padding-right: 2px;
                  }
                  & .ms-TextField-field {
                    min-width: auto;
                    padding-bottom: 5px;
                    padding-left: 5px;
                    padding-right: 5px;
                    padding-top: 5px;
                    text-overflow: clip;
                  }
            >
              <div
                className=
                    ms-TextField-wrapper

              >
                <div
                  className=
                      ms-TextField-fieldGroup
                      {
                        align-items: stretch;
                        background: #ffffff;
                        border: 1px solid #a6a6a6;
                        box-shadow: none;
                        box-sizing: border-box;
                        display: flex;
                        flex-direction: row;
                        height: 32px;
                        margin-bottom: 0px;
                        margin-left: 0px;
                        margin-right: 0px;
                        margin-top: 0px;
                        padding-bottom: 0px;
                        padding-left: 0px;
                        padding-right: 0px;
                        padding-top: 0px;
                        position: relative;
                      }
                      &:hover {
                        border-color: #333333;
                      }
                      @media screen and (-ms-high-contrast: active){&:hover {
                        border-color: Highlight;
                      }
                >
                  <input
                    aria-invalid={false}
                    aria-label="Blue"
                    className=
                        ms-TextField-field
                        {
                          background-color: transparent;
                          background: none;
                          border-radius: 0px;
                          border: none;
                          box-shadow: none;
                          box-sizing: border-box;
                          color: #333333;
                          font-size: 14px;
                          margin-bottom: 0px;
                          margin-left: 0px;
                          margin-right: 0px;
                          margin-top: 0px;
                          min-width: 0px;
                          outline: 0px;
                          padding-bottom: 0;
                          padding-left: 12px;
                          padding-right: 12px;
                          padding-top: 0;
                          text-overflow: ellipsis;
                          width: 100%;
                        }
                        &:active, &:focus, &:hover {
                          outline: 0px;
                        }
                        &::-ms-clear {
                          display: none;
                        }
                        &::placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                        &:-ms-input-placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                    id="TextField6"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    spellCheck={false}
                    type="text"
                    value="255"
                  />
                </div>
              </div>
            </div>
          </td>
          <td
            style={
              Object {
                "width": "18%",
              }
            }
          >
            <div
              className=
                  ms-TextField
                  ms-ColorPicker-input
                  {
                    border: none;
                    box-shadow: none;
                    box-sizing: border-box;
                    height: 30px;
                    margin-bottom: 0px;
                    margin-left: 0px;
                    margin-right: 0px;
                    margin-top: 0px;
                    padding-bottom: 0px;
                    padding-left: 0px;
                    padding-right: 0px;
                    padding-top: 0px;
                    position: relative;
                    width: 100%;
                  }
                  @media screen and (-ms-high-contrast: active){& {
                    border-width: 2px;
                  }
                  &.ms-TextField {
                    padding-right: 2px;
                  }
                  & .ms-TextField-field {
                    min-width: auto;
                    padding-bottom: 5px;
                    padding-left: 5px;
                    padding-right: 5px;
                    padding-top: 5px;
                    text-overflow: clip;
                  }
            >
              <div
                className=
                    ms-TextField-wrapper

              >
                <div
                  className=
                      ms-TextField-fieldGroup
                      {
                        align-items: stretch;
                        background: #ffffff;
                        border: 1px solid #a6a6a6;
                        box-shadow: none;
                        box-sizing: border-box;
                        display: flex;
                        flex-direction: row;
                        height: 32px;
                        margin-bottom: 0px;
                        margin-left: 0px;
                        margin-right: 0px;
                        margin-top: 0px;
                        padding-bottom: 0px;
                        padding-left: 0px;
                        padding-right: 0px;
                        padding-top: 0px;
                        position: relative;
                      }
                      &:hover {
                        border-color: #333333;
                      }
                      @media screen and (-ms-high-contrast: active){&:hover {
                        border-color: Highlight;
                      }
                >
                  <input
                    aria-invalid={false}
                    aria-label="Alpha"
                    className=
                        ms-TextField-field
                        {
                          background-color: transparent;
                          background: none;
                          border-radius: 0px;
                          border: none;
                          box-shadow: none;
                          box-sizing: border-box;
                          color: #333333;
                          font-size: 14px;
                          margin-bottom: 0px;
                          margin-left: 0px;
                          margin-right: 0px;
                          margin-top: 0px;
                          min-width: 0px;
                          outline: 0px;
                          padding-bottom: 0;
                          padding-left: 12px;
                          padding-right: 12px;
                          padding-top: 0;
                          text-overflow: ellipsis;
                          width: 100%;
                        }
                        &:active, &:focus, &:hover {
                          outline: 0px;
                        }
                        &::-ms-clear {
                          display: none;
                        }
                        &::placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                        &:-ms-input-placeholder {
                          color: #666666;
                          opacity: 1;
                        }
                    id="TextField8"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    spellCheck={false}
                    type="text"
                    value="100"
                  />
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
